### ISSUE TITLE
Support for prepared dynamic statements(LRU Statement Cache Size)

### ISSUE THREAD

**Issue by steve-chavez** [nikita-volkov/hasql-dynamic-statements#3]
Hi @nikita-volkov.

For performance reasons, I'm trying to use prepared statements on PostgREST.

While trying to use `hasql-dynamic-statements`, I've noticed that it [doesn't prepare](https://github.com/nikita-volkov/hasql-dynamic-statements/blob/f31967526e603253b0fd4a68e5c6af18b02c34a8/library/Hasql/DynamicStatements/Statement.hs#L64)(flag set to False) the statements.

On https://github.com/nikita-volkov/hasql/pull/111#issuecomment-492143736, you mention:

> don't forget that you'll have to set the "prepared" flag to False, because otherwise you'll overflow the prepared statements cache on the Postgres side.

I understand the problem about overflowing the cache size. However, wouldn't it be possible to define a `Statement Cache Size` on Hasql and LRU the overflowing statements?(1)

I'm wondering if you'd be interested in supporting this use case.

(1): I've noted that JDBC has a [similar way of handling its statement cache size](https://docs.oracle.com/middleware/1213/wls/WLACH/taskhelp/jdbc/jdbc_datasources/ConfigureTheStatementCacheForADataSource.html).  This cache size is actually the number of cached statements.

**Comment by steve-chavez** [comment]
(PreparedStatementRegistry already has a [counter](https://github.com/nikita-volkov/hasql/blob/fdb58ec30eadc263ac7800e0d7b1a1d8ae3165a3/library/Hasql/Private/PreparedStatementRegistry.hs#L25). Perhaps that could be reused for limiting the cache size.)

**Comment by nikita-volkov** [comment]
I see several problems with this.

1. There is a performance cost:
 
    - It will introduce overhead of looking up statements in the registry on every statement execution and overhead of managing LRU. Right now unprepared statements avoid all that.
    - It is not simply about modifying the client-side registry. We'll also need to implement a mechanism that will invalidate particular prepared statements on the server-side. And that will introduce traffic overhead.

2. There is no point in any caching of statements that are truly dynamic. I'm talking about the kind of statements that is likely to get a different SQL on each execution. More so, it's likely to produce a large invalidation traffic.

3. In many cases the benefit of prepared statements is negligible, so there's a risk that the overheads introduced by such solution may exceed the gains.

However I sense that I understand your problem and I think I see a potential solution.

There are cases when the dynamicity of a statement is of limits known ahead. IOW, there's a limited amount combinations of how that kind of statement can arrange its SQL. For these cases it makes sense to allow the user to state that he wants a dynamic statement cached. So I'm open to exposing that flag.

Also reminding just in case you need it. You already have the control over invalidation of cache: by reconnecting you get all caches reset.

**Comment by steve-chavez** [comment]
> For these cases it makes sense to allow the user to state that he wants a dynamic statement cached. So I'm open to exposing that flag.

@nikita-volkov Thank you! That would definitely help.

For now, thanks to your [example on Hackage](http://hackage.haskell.org/package/hasql-dynamic-statements-0.2.0.3/docs/Hasql-DynamicStatements-Statement.html), I've found that I can add dynamic parameters on Hasql by doing:

```haskell
query :: [ByteString] -> Statement () Results
query inputs = Statement sql encoder decoder True
where
  encoder = mconcat $ (Encoders.param . ($<) Encoders.unknown) <$> inputs
  decoder = ...
  sql = ...
```

(All of PostgREST's parameters use the `unknown` decoder, so no problem there)

> There are cases when the dynamicity of a statement is of limits known ahead. IOW, there's a limited amount combinations of how that kind of statement can arrange its SQL

Yes, that's true for us as well. However, having a limited statement cache size would give us an additional safeguard, in case some client gets creative and manages to abuse producing prepared statements.

> It will introduce overhead of looking up statements in the registry on every statement execution and overhead of managing LRU. Right now unprepared statements avoid all that.

How about having a **Fixed** Statement Cache Size? From the [JDBC doc](https://docs.oracle.com/middleware/1213/wls/WLACH/taskhelp/jdbc/jdbc_datasources/ConfigureTheStatementCacheForADataSource.html):

**"No new statements are cached unless the cache is manually cleared or the cache size is increased"**

I'm not sure of the implementation, but would that be possible on Hasql without causing much overhead? It looks like it could be done by having a `max` on the PreparedStatementRegistry and not allowing the `counter` to exceed that.

> In many cases the benefit of prepared statements is negligible

Definitely PostgREST specific, but I've noticed around 20% more throughput when switching to prepared statements. The gains would be huge for us(specially now that there's some talk about getting into techempower benchmarks).
